server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -435,6 +467,15 @@ private ThreadContextStruct putResponseHeaders(Map<String, List<String>> headers
 
         private ThreadContextStruct putResponse(final String key, final String value, final Function<String, String> uniqueValue) {
             assert value != null;
+            long curWarningHeaderSize = 0;
+            if (key.equals("Warning")) {
+                // check if we can add another warning header, if max count or size exceeded
+                final int warningHeaderCount = this.responseHeaders.containsKey("Warning")? this.responseHeaders.get("Warning").size() : 0;
+                if (warningHeaderCount >= maxWarningHeaderCount) return this; //can't add max count reached
```

Comment:
I think it'd be useful to get one more warning when the limit is hit, saying that there were more warnings but we dropped them because `http.max_warning_header_count` is set to `<n>`, and similarly for the size limit.

Reply:
This is a great suggestion, I think it should be a warning in the server logs as opposed to an additional warning header that eats into our budgets.

Reply:
I think it's more useful if we spend the last of the budget saying the list is truncated, in anticipation of a time when someone forgets that the list they're looking at might be truncated and makes a poor decision on the assumption that it's complete. These excessive warnings are often repetitive, so the specifics of the 62nd warning seem less valuable to me.

(Not that it shouldn't be in the server logs too.)

Reply:
My concern here is if a user sets the budget to `H` headers and `B` bytes because they can not handle more than that (e.g., the common case being a proxy) then we have to subtract a header (or possibly many) to stay under the `(H, B)` budget after we include the missed warnings warning.